userspace/libsinsp/parsers.cpp

```diff
@@ -5232,13 +5221,41 @@ void sinsp_parser::parse_container_json_evt(sinsp_evt *evt)
 			}
 		}
 
+		evt->set_tinfo_ref(container_info->get_tinfo(m_inspector));
+		evt->set_tinfo(evt->get_tinfo_ref().get());
+		container_info->set_lookup_status(static_cast<sinsp_container_lookup::state>(lookup_state.asUInt()));
+		switch(container_info->get_lookup_status())
+		{
+		// Preserve case where lookup_state == STARTED from scap file was set to FAILED
+		// or otherwise set to false.
+		case sinsp_container_lookup::state::FAILED:
+			break;
+		default:
+			{
+				// note: The container image is the most crucial field from a security incident response perspective. 
+				// We aim to raise the bar for successful container lookups. Conversely, pod sandboxes do not include  
+				// a container image in the API response.
+				if (!container_info->m_image.empty() || container_info->is_pod_sandbox())
+				{
+					// Marked as SUCCESSFUL if and only if the container image field present or 
+					// if it's a pod sandbox container
+					container_info->set_lookup_status(sinsp_container_lookup::state::SUCCESSFUL);
+				} else
+				{
+					container_info->set_lookup_status(sinsp_container_lookup::state::FAILED);
+				}
```

Comment:
Sorry, but I disagree. This is hardcoding your policy ("all containers must have an image name"), which looks sensible but https://xkcd.com/1172/ .

How is the FAILED state an improvement here? ->this<- is the metadata we have, if it contains an empty image name, so be it, no reason to pretend all the other metadata does not exist. Given how fluid the CRI spec seems to be, I can totally imagine `some-new-fancy-wrapper-over-runc run sha256:abcdef` to not expose the image name at all (since we don't have it), but still, e.g. the labels might be valuable.

Not to mention, this is generic code, for all container engines, some of which may not even have a notion of image names (I'm not sure e.g. LXC does). If I am in the minority and we decide to change the policy to require an image name, please at the very least place it directly in the container engine(s).

I understand you need the image name for anything interesting, but the real fix would be to make metadata queries fast and reliable (which we know is impossible in all cases but we can keep trying). This is just shooting the messenger: given how the docker/cri engines are written, if you get a success without an image name, I'd assume we have not received an image name from the API and there's nothing we can do about it. OK, maybe retrying the query a bit later might help (doubt it but stranger things have happened), but that's entirely the business of a particular engine.

Reply:
Jumping in late in this discussion. From my high-level understanding, this seem we're deciding between:

- Discarding all partially-complete metadata in case the container image is not available, and force new fetch attempt later on thus consuming extra CPU in the hope of getting more complete data
- Consider the fetch successful even in case some data (such as container image) is missing

Both comes with evident pros and cons. If we struggle finding an agreement, to me this feels like the kind of behavior that should be configurable and not be enforced on the libs adopters. Of course I agree that security-related use cases have top-priority overall for the Falco project, but I also see that the libs themselves are an arsenal of system inspection tools that can be the foundation of different projects and products as well. To me, this feels like something we could negotiate as a "retry policy" new configuration point. Just my two cents, I trust both of you with your own expertise and evaluations.

> Not to mention, this is generic code, for all container engines, some of which may not even have a notion of image names (I'm not sure e.g. LXC does). If I am in the minority and we decide to change the policy to require an image name, please at the very least place it directly in the container engine(s).

This one I'm very onboard with -- I think this should not affect the behavior of other container runtimes.

> retrying later (for some definition of later, probably other than "the next syscall"), but that has to be done in an explicit manner.

Agree also on this, this "retry policy" definitely has to happen in a controlled or throttled way.

Reply:
Given that we have the new ticket https://github.com/falcosecurity/libs/issues/1708, let's hold off on any changes in this regard in the parser, and someone shall refactor it in the future. Please be aware of https://github.com/falcosecurity/libs/pull/1707. It's incorrect to initialize a lookup to SUCCESSFUL, by the way. I believe this is the strongest indication of a very suboptimal shared design across the container engines we support.